File: RecursionSolutions/ruby/power_set.rb
# Recursive Power Set Solution

def powerSet(set)
  return [set] if set.empty?
  p = set.pop
  subset = powerSet(set)  
  subset | subset.map { |x| x | [p] }
end

p powerSet(['a', 'b', 'c', 'd'])

#solution without the confusing |

def powerSet(array)
  return [[]] if array.empty?
  set = powerset(array[1..-1])
  set.map{|i| [array.first] + i} + set
end
p powerSet(['a', 'b', 'c', 'd'])

File: solutions/RecursionSolutions/js/powerSet.js
// Recursive Power Set Solution

var powerset = function(arr, prefix, subsets) {
  subsets = subsets || [];
  prefix  = prefix  || [];
  if (arr.length) {
    powerset(arr.slice(1), prefix.concat(arr[0]), subsets)
    powerset(arr.slice(1), prefix, subsets)
  } else {
    subsets.push(prefix)
  }
  return subsets
};

// console.log(powerset('abcd'));
console.log(powerset('abc'))
